packages/core/src/lib/history.ts

```diff
@@ -0,0 +1,110 @@
+import { LogResult, simpleGit } from 'simple-git';
+import {
+  CoreConfig,
+  Format,
+  PersistConfig,
+  UploadConfig,
+  uploadConfigSchema,
+} from '@code-pushup/models';
+import { getCurrentBranchOrTag, safeCheckout } from '@code-pushup/utils';
+import { collectAndPersistReports } from './collect-and-persist';
+import { GlobalOptions } from './types';
+import { upload as uploadCommandLogic } from './upload';
+
+export type HistoryOnlyOptions = {
+  targetBranch?: string;
+  uploadReports?: boolean;
+  forceCleanStatus?: true;
+};
+export type HistoryOptions = Required<
+  Pick<CoreConfig, 'plugins' | 'categories'> & {
+    persist: Required<PersistConfig>;
+  } & { upload: Required<UploadConfig> }
+> &
+  GlobalOptions &
+  HistoryOnlyOptions;
+
+export async function history(
+  config: HistoryOptions,
+  commits: string[],
+): Promise<string[]> {
+  const initialBranch: string = await getCurrentBranchOrTag();
+
+  const { uploadReports = true } = config as unknown as HistoryOptions;
+  if (!uploadReports) {
+    console.warn('Upload is skipped because uploadReports is set to false');
+  }
+
+  const reports: string[] = [];
+  // eslint-disable-next-line functional/no-loop-statements
+  for (const commit of commits) {
+    console.info(`Collect ${commit}`);
+    await safeCheckout(commit, { forceCleanStatus: config.forceCleanStatus });
+
+    const currentConfig: HistoryOptions = {
+      ...config,
+      persist: {
+        ...config.persist,
+        format: ['json' as Format],
+        filename: `${commit}-report`,
+      },
+    };
+
+    await collectAndPersistReports(currentConfig);
+
+    if (uploadReports) {
+      const result = uploadConfigSchema.safeParse(currentConfig.upload);
+      if (result.success) {
+        await uploadCommandLogic({ ...currentConfig, upload: result.data });
+      } else {
+        console.error(`Collecting ${commit} failed.`);
+        console.error(result.error);
+      }
```

Comment:
The error message doesn't mention the cause - invalid upload config. Can that even happen, though? If this is to be triggered from the command-line, then it will have already been parsed in the CLI middleware. So probably the `safeParse` can be ommitted here.

```suggestion
      await uploadCommandLogic(currentConfig);
```

Reply:
upload it self could be undefined as it is optional in core config. I will adjust the typing of `HistoryCliOptions` to be in line with CoreConfig.

Reply:
Done! I also adjusted the log message accordingly.